sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs

```diff
@@ -266,141 +264,147 @@ public virtual async Task<Response<UploadLogsResult>> UploadAsync<T>(string rule
+            foreach (BatchedLogs<T> batch in Batch(logs, options))
             {
-                // A list of tasks that are currently executing which will
-                // always be smaller than or equal to MaxWorkerCount
-                List<Task<BatchUpload>> runningTasks = new();
-                // Partition the stream into individual blocks
-                foreach (BatchedLogs<T> batch in Batch(logs, options))
+                try
                 {
                     // Start staging the next batch (but don't await the Task!)
-                    Task<BatchUpload> task = CommitBatchListSyncOrAsync(
+                    Task<Response> task = UploadBatchListSyncOrAsync(
                         batch,
                         ruleId,
                         streamName,
-                        true,
+                        async: true,
                         cancellationToken);
 
                     // Add the block to our task and commit lists
-                    runningTasks.Add(task);
+                    runningTasks.Add((task, batch.LogsList.Count));
 
                     // If we run out of workers
                     if (runningTasks.Count >= _maxWorkerCount)
                     {
                         // Wait for at least one of them to finish
-                        await Task.WhenAny(runningTasks).ConfigureAwait(false);
+                        await Task.WhenAny(runningTasks.Select(_ => _.CurrentTask)).ConfigureAwait(false);
                         // Clear any completed blocks from the task list
                         for (int i = 0; i < runningTasks.Count; i++)
                         {
-                            Task<BatchUpload> runningTask = runningTasks[i];
+                            Task<Response> runningTask = runningTasks[i].CurrentTask;
                             if (!runningTask.IsCompleted)
                             {
                                 continue;
                             }
-
-                            // Before removing runningTask - check if it has any errors and add to error list
-                            if (runningTask.Result.Error != null)
-                                errors.Add(runningTask.Result.Error);
-
+                            // If current task has an exception, log the exception and add number of logs in this task to failed logs
+                            if (runningTask.Exception != null)
+                            {
+                                exceptions = AddException(
+                                    exceptions,
+                                    runningTask.Exception);
+                                logsFailed += runningTasks[i].LogsCount;
+                            }
+                            // If current task returned a response that was not a success, log the exception and add number of logs in this task to failed logs
+                            if (runningTask.Result.Status != 204)
```

Comment:
Should this be an else if, like it is here - https://github.com/Azure/azure-sdk-for-net/pull/31846/files#diff-6767d7871ae91cf5eb10fa3e4f24a9d5bbd9f9dbcc22bcbd72be0eabe315c103R358?

Can we extract this into a method, e.g. ProcessCompletedTasks?